README.md
# WShot

A command line app to take screenshots of web pages, written in NodeJS.

## Install

Run `npm install -g wshot`

## How to use

You can get a screenshot of a URL (must start with http:// or https://)

`wshot https://www.gov.uk`

Or multiple URLs, using a text file with each URL on a new line:

`wshot urls.txt`

Screenshots are saved into a folder called `screenshots`.

You can set browser width, it defaults to 1200 pixels.

`wshot https://www.gov.uk --width 320`

By default, you'll get a screenshot of the full height of the page. To get a set height:

`wshot https://www.gov.uk --height 800`


CHANGELOG.md
# Changelog

## 1.1.1

Fixes:

 - [Handle timeout and bad URL errors](https://github.com/joelanman/wshot/pull/13)
 - [Improve help description](https://github.com/joelanman/wshot/pull/12)
 - [Check if input is missing, add name and examples for commander help](https://github.com/joelanman/wshot/pull/9)

## 1.1.0

New features:

 - [Turn off fullPage automatically if height is set](https://github.com/joelanman/wshot/pull/5)
 - [Use babel to support node 6 (4)](https://github.com/joelanman/wshot/pull/4)
 - [Use Standard JS style guide (3)](https://github.com/joelanman/wshot/pull/3)


screenshot-cli.js
#!/usr/bin/env node
const puppeteer = require('puppeteer')
const fs = require('fs')
const program = require('commander')
const screenshotFolder = process.cwd() + '/screenshots'

if (!fs.existsSync(screenshotFolder)){
    fs.mkdirSync(screenshotFolder)
}

program
  .arguments('<url>')
  .option('-w, --width <width>', 'The width of the viewport in pixels (defaults to 1200)')
  .option('-h, --height <height>', 'The height of the viewport in pixels (defaults to 1200)')
  .option('-f, --fullPage <fullPage>', 'Full page (defaults to true)')
  .action(async function(url) {
    const browser = await puppeteer.launch()
    const page = await browser.newPage()
    await page.setViewport({
        width: parseInt(program.width || 1200),
        height: parseInt(program.height || 1200)
    });
    await page.goto(url)
    let filename = url.replace(/^https?:\/\//,'')
    filename = filename.replace(/^www\./,'')
    filename = filename.replace(/\//g,'-')+'.png'
    await page.screenshot({
        fullPage: program.fullPage !== 'false',
        path: screenshotFolder + '/' + filename
    })
    await browser.close()
    return
  })
  .parse(process.argv);


src/screenshot.js
#!/usr/bin/env node
const puppeteer = require('puppeteer')
const fs = require('fs')
const program = require('commander')
const screenshotFolder = process.cwd() + '/screenshots'

if (!fs.existsSync(screenshotFolder)) {
  fs.mkdirSync(screenshotFolder)
}

function delay (timeout) {
  return new Promise((resolve) => {
    setTimeout(resolve, timeout)
  })
}

program
  .arguments('<input>')
  .option('-w, --width <width>', 'The width of the viewport in pixels (defaults to 1200)')
  .option('-h, --height <height>', 'The height of the viewport in pixels (defaults to the full height of the page)')
  .action(async function (input) {

    const fullPage = (program.height === undefined)

    let urls = []

    try {
      // read file of URLs
      const urlsFile = fs.readFileSync(input, 'utf8')
      urls = urlsFile.split('\n')
    } catch (e) {
      // not a file, probably a URL
      urls.push(input)
    }

    const browser = await puppeteer.launch()
    const page = await browser.newPage()
    await page.setViewport({
      width: parseInt(program.width || 1200),
      height: parseInt(program.height || 1200)
    })
    for (let index = 0; index < urls.length; index++) {
      let url = urls[index]
      if (url === '') {
        continue
      }
      console.log(url)
      await page.goto(url)
      let filename = url.replace(/^https?:\/\//, '')
      filename = filename.replace(/^www\./, '')
      filename = filename.replace(/\//g, '-') + '.png'
      await page.screenshot({
        fullPage: fullPage,
        path: screenshotFolder + '/' + filename
      })
      if (index < urls.length - 1) {
        await delay(1000)
      }
    }
    await browser.close()
  })
  .parse(process.argv)


src/wshot.js
#!/usr/bin/env node
'use strict'

const puppeteer = require('puppeteer')
const fs = require('fs')
const program = require('commander')
const screenshotFolder = process.cwd() + '/screenshots'

if (!fs.existsSync(screenshotFolder)) {
  fs.mkdirSync(screenshotFolder)
}

function delay (timeout) {
  return new Promise((resolve) => {
    setTimeout(resolve, timeout)
  })
}

const description = `Take screenshots of web pages.

  Examples:

  wshot https://www.gov.uk
  wshot myURLs.txt

  If you supply a text file, it should be a list of URLs, each on a new line.
  wshot will take a screenshot of each URL.
`

program
  .arguments('<input>')
  .description(description)
  .option('-w, --width <width>', 'The width of the viewport in pixels (defaults to 1200)')
  .option('-h, --height <height>', 'The height of the viewport in pixels (defaults to the full height of the page)')
  .action(function (input) {
     program.input = input
  })
  .parse(process.argv)

if (program.input === undefined){
  console.error("Missing input, try wshot --help")
  process.exit()
}

const input = program.input

const fullPage = (program.height === undefined)

let urls = []

try {
  // read file of URLs
  const urlsFile = fs.readFileSync(input, 'utf8')
  urls = urlsFile.split('\n')
} catch (e) {
  // not a file, probably a URL
  urls.push(input)
}

(async () => {

  const timeout = 10 * 1000
  const browser = await puppeteer.launch()
  const page = await browser.newPage()
  await page.setViewport({
    width: parseInt(program.width || 1200),
    height: parseInt(program.height || 1200)
  })
  for (let index = 0; index < urls.length; index++) {
    let url = urls[index]
    if (url === '') {
      continue
    }
    console.log(url)
    try {
      await page.goto(url, {
        timeout: timeout
      })
    } catch (error){
      if (error.message.indexOf("Navigation Timeout Exceeded") === 0){
        console.error(`Error: Server did not respond within ${timeout / 1000} seconds`)
        continue
      } else if (error.message.indexOf("Protocol error (Page.navigate): Cannot navigate to invalid URL") === 0){
        console.error(`Error: This doesn't seem to be a correct URL`)
        continue
      } else {
        console.error(error)
        continue
      }
    }
    let filename = url.replace(/^https?:\/\//, '')
    filename = filename.replace(/^www\./, '')
    filename = filename.replace(/\//g, '-') + '.png'
    await page.screenshot({
      fullPage: fullPage,
      path: screenshotFolder + '/' + filename
    })
    if (index < urls.length - 1) {
      await delay(1000)
    }
  }
  await browser.close()

})()
